Introduce the DirtyRect type

nical opened the pull request:
This is Alexis' [DirtyRect patch](https://phabricator.services.mozilla.com/D9559) completed and rebased on top of #3272.

It turns `Option<Rect>` into an explicit type which is a lot easier to understand as `None` previously meant different things in different areas of the code.

nical commented:
Try push: https://treeherder.mozilla.org/#/jobs?repo=try&revision=e3bde69a576c86d71a88e80c8c77b4a295562642 I added some fixes to get wrench to build after the try push but the gecko visible parts are in there.

Gankro commented:
SGTM, but I wrote most of the second commit so I'm not sure if my vote counts 😅

(signed off on the first one in the other PR)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #3282) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #3272) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #3284) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #3321) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #3326) made this pull request unmergeable. Please resolve the merge conflicts.

nical commented:
New try push with the latest changes https://treeherder.mozilla.org/#/jobs?repo=try&revision=f91e139cb384a0f827ce46d8a69f7f9425e466f2

kvark commented:
There are R5 failures in the try.

nical commented:
> There are R5 failures in the try.

These are fuzzable and appeared after I rebased the webrender parts so I suspect they come from another commit that doesn't have the corresponding gecko change in the push. There is also an unexpected pas that I can't take credit for.
I just rebased the gecko parts again, I'll land this if the next try push comes back green.

nical commented:
@bors-servo r=kvark

bors-servo commented:
:pushpin: Commit b18556e has been approved by `kvark`

<!-- @bors-servo r=kvark b18556ea162a84e5e30c89b79b2c5121478aaadc -->

bors-servo commented:
:hourglass: Testing commit b18556ea162a84e5e30c89b79b2c5121478aaadc with merge af2b372624db589115511b4705849a33e6acd35d...
